test leak in TestInfo

FAIL:TestInfo

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/tidb_ghpr_unit_test/runs/24373/nodes/84/log/?start=0
```
[2020-02-17T03:08:06.587Z] === RUN   TestInfo
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.308 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.TestInfo\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain_test.go:150\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.311 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.314 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.318 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.321 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.324 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.327 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.587Z] [2020/02/17 11:08:01.330 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] [2020/02/17 11:08:01.334 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] [2020/02/17 11:08:01.337 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] [2020/02/17 11:08:01.340 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] [2020/02/17 11:08:01.343 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] [2020/02/17 11:08:01.346 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] [2020/02/17 11:08:01.349 +08:00] [ERROR] [syncer.go:199] ["close session failed"] [] [stack="github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Done\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/ddl/util/syncer.go:199\ngithub.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop\n\t/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:485"]
[2020-02-17T03:08:06.588Z] {"level":"warn","ts":"2020-02-17T11:08:01.350+0800","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-516b5633-69d2-49d2-a3a5-8ed7da21c073/localhost:1525069566242647020","attempt":0,"error":"rpc error: code = NotFound desc = etcdserver: requested lease not found"}
[2020-02-17T03:08:06.588Z] --- FAIL: TestInfo (2.75s)
[2020-02-17T03:08:06.588Z]     leaktest.go:143: Test TestInfo check-count 50 appears to have leaked: github.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop(0xc0006026c0, 0x2faf080)
[2020-02-17T03:08:06.588Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:469 +0x269
[2020-02-17T03:08:06.588Z]         created by github.com/pingcap/tidb/domain.(*Domain).Init
[2020-02-17T03:08:06.588Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test@2/go/src/github.com/pingcap/tidb/domain/domain.go:696 +0x59c
[2020-02-17T03:08:06.588Z] FAIL
```

2. What did you expect to see?



3. What did you see instead?



4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

Another error message.
```
[2020-03-16T05:13:27.940Z] --- FAIL: TestInfo (2.74s)
[2020-03-16T05:13:27.940Z]     leaktest.go:143: Test TestInfo check-count 50 appears to have leaked: go.etcd.io/etcd/clientv3.waitRetryBackoff(0x2f8ff00, 0xc000511a40, 0x5d, 0xc000ae0660, 0x3, 0x3)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/clientv3/retry_interceptor.go:267 +0xfd
[2020-03-16T05:13:27.940Z]         go.etcd.io/etcd/clientv3.(*Client).unaryClientInterceptor.func1(0x2f8ff00, 0xc000511a40, 0x2bddb89, 0x1f, 0x2a127a0, 0xc001a331a0, 0x29ff480, 0xc00167c1e8, 0xc0004d7180, 0x2c64820, ...)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/clientv3/retry_interceptor.go:49 +0x14f
[2020-03-16T05:13:27.940Z]         google.golang.org/grpc.(*ClientConn).Invoke(0xc0004d7180, 0x2f8ff00, 0xc000511a40, 0x2bddb89, 0x1f, 0x2a127a0, 0xc001a331a0, 0x29ff480, 0xc00167c1e8, 0xc000511aa0, ...)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/google.golang.org/grpc@v1.25.1/call.go:35 +0x109
[2020-03-16T05:13:27.940Z]         google.golang.org/grpc.Invoke(...)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/google.golang.org/grpc@v1.25.1/call.go:60
[2020-03-16T05:13:27.940Z]         go.etcd.io/etcd/etcdserver/etcdserverpb.(*leaseClient).LeaseRevoke(0xc0002dcbd0, 0x2f8ff00, 0xc000511a40, 0xc001a331a0, 0xc000511aa0, 0x4, 0x6, 0x6, 0xf7b4c8, 0x10)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/etcdserverpb/rpc.pb.go:3801 +0xcf
[2020-03-16T05:13:27.940Z]         go.etcd.io/etcd/clientv3.(*retryLeaseClient).LeaseRevoke(0xc0020744c0, 0x2f8ff00, 0xc000511a40, 0xc001a331a0, 0x47dec00, 0x3, 0x3, 0x2f8ff00, 0xc000511a40, 0xc001949fe0)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/clientv3/retry.go:148 +0xeb
[2020-03-16T05:13:27.940Z]         go.etcd.io/etcd/clientv3.(*lessor).Revoke(0xc000c83cc0, 0x2f8ff00, 0xc000511a40, 0x3feb70e1c1edc705, 0xc000511a40, 0xc001949fe0, 0x19)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/clientv3/lease.go:231 +0x92
[2020-03-16T05:13:27.940Z]         github.com/pingcap/tidb/owner.(*ownerManager).revokeSession(0xc000c45c70, 0xc0002bb6d0, 0x4a, 0x3feb70e1c1edc705)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/owner/manager.go:290 +0xaa
[2020-03-16T05:13:27.940Z]         github.com/pingcap/tidb/owner.(*ownerManager).campaignLoop(0xc000c45c70, 0x2f8fe80, 0xc001888a80, 0xc00120f020)
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/owner/manager.go:244 +0x10b7
[2020-03-16T05:13:27.940Z]         created by github.com/pingcap/tidb/owner.(*ownerManager).CampaignOwner
[2020-03-16T05:13:27.940Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/owner/manager.go:186 +0x1b0
[2020-03-16T05:13:27.940Z] FAIL
[2020-03-16T05:13:27.940Z] FAIL	github.com/pingcap/tidb/domain	3.185s
```
https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_unit_test/detail/tidb_ghpr_unit_test/27672/pipeline

Seen at https://github.com/pingcap/tidb/pull/15194#issuecomment-599348928

--- a/domain/db_test.go
+++ b/domain/db_test.go
@@ -20,6 +20,7 @@ import (
 	. "github.com/pingcap/check"
 	"github.com/pingcap/tidb/session"
 	"github.com/pingcap/tidb/store/mockstore"
+	"github.com/pingcap/tidb/util/testleak"
 )
 
 type dbTestSuite struct{}
@@ -27,13 +28,17 @@ type dbTestSuite struct{}
 var _ = Suite(&dbTestSuite{})
 
 func (ts *dbTestSuite) TestIntegration(c *C) {
+	testleak.BeforeTest()
+	defer testleak.AfterTest(c)()
 	var err error
 	lease := 50 * time.Millisecond
 	store, err := mockstore.NewMockStore()
 	c.Assert(err, IsNil)
+	defer store.Close()
 	session.SetSchemaLease(lease)
-	_, err = session.BootstrapSession(store)
+	domain, err := session.BootstrapSession(store)
 	c.Assert(err, IsNil)
+	defer domain.Close()
 
 	// for NotifyUpdatePrivilege
 	createRoleSQL := `CREATE ROLE 'test'@'localhost';`

--- a/domain/domain_test.go
+++ b/domain/domain_test.go
@@ -18,9 +18,6 @@ import (
 	"crypto/tls"
 	"math"
 	"net"
-	"runtime"
-	"sort"
-	"strings"
 	"testing"
 	"time"
 
@@ -40,12 +37,10 @@ import (
 	"github.com/pingcap/tidb/store/mockstore"
 	"github.com/pingcap/tidb/store/tikv/oracle"
 	"github.com/pingcap/tidb/util"
-	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tidb/util/mock"
 	"github.com/pingcap/tidb/util/testleak"
 	dto "github.com/prometheus/client_model/go"
 	"go.etcd.io/etcd/integration"
-	"go.uber.org/zap"
 )
 
 func TestT(t *testing.T) {
@@ -91,137 +86,12 @@ func unixSocketAvailable() bool {
 	return false
 }
 
-// For debug only, will be removed later.
-// TODO: remove these tests for investigating the leak problem.
-func interestingGoroutines() (gs []string) {
-	buf := make([]byte, 2<<20)
-	buf = buf[:runtime.Stack(buf, true)]
-	for _, g := range strings.Split(string(buf), "\n\n") {
-		sl := strings.SplitN(g, "\n", 2)
-		if len(sl) != 2 {
-			continue
-		}
-		stack := strings.TrimSpace(sl[1])
-		if stack == "" ||
-			strings.Contains(stack, "created by github.com/pingcap/tidb.init") ||
-			strings.Contains(stack, "testing.RunTests") ||
-			strings.Contains(stack, "check.(*resultTracker).start") ||
-			strings.Contains(stack, "check.(*suiteRunner).runFunc") ||
-			strings.Contains(stack, "check.(*suiteRunner).parallelRun") ||
-			strings.Contains(stack, "localstore.(*dbStore).scheduler") ||
-			strings.Contains(stack, "tikv.(*noGCHandler).Start") ||
-			strings.Contains(stack, "ddl.(*ddl).start") ||
-			strings.Contains(stack, "ddl.(*delRange).startEmulator") ||
-			strings.Contains(stack, "domain.NewDomain") ||
-			strings.Contains(stack, "testing.(*T).Run") ||
-			strings.Contains(stack, "domain.(*Domain).LoadPrivilegeLoop") ||
-			strings.Contains(stack, "domain.(*Domain).UpdateTableStatsLoop") ||
-			strings.Contains(stack, "testing.Main(") ||
-			strings.Contains(stack, "runtime.goexit") ||
-			strings.Contains(stack, "created by runtime.gc") ||
-			strings.Contains(stack, "interestingGoroutines") ||
-			strings.Contains(stack, "runtime.MHeap_Scavenger") ||
-			// these go routines are async terminated, so they may still alive after test end, thus cause
-			// false positive leak failures
-			strings.Contains(stack, "google.golang.org/grpc.(*addrConn).resetTransport") ||
-			strings.Contains(stack, "google.golang.org/grpc.(*ccBalancerWrapper).watcher") ||
-			strings.Contains(stack, "github.com/pingcap/goleveldb/leveldb/util.(*BufferPool).drain") ||
-			strings.Contains(stack, "github.com/pingcap/goleveldb/leveldb.(*DB).compactionError") ||
-			strings.Contains(stack, "github.com/pingcap/goleveldb/leveldb.(*DB).mpoolDrain") {
-			continue
-		}
-		gs = append(gs, stack)
-	}
-	sort.Strings(gs)
-	return
-}
-
-var beforeTestGoroutines = map[string]bool{}
-var testGoroutinesInited bool
-
-// BeforeTest gets the current goroutines.
-// It's used for check.Suite.SetUpSuite() function.
-// Now it's only used in the tidb_test.go.
-func BeforeTest() {
-	for _, g := range interestingGoroutines() {
-		beforeTestGoroutines[g] = true
-	}
-	testGoroutinesInited = true
-}
-
-const defaultCheckCnt = 50
-
-func checkLeakAfterTest(errorFunc func(cnt int, g string)) func() {
-	// After `BeforeTest`, `beforeTestGoroutines` may still be empty, in this case,
-	// we shouldn't init it again.
-	if !testGoroutinesInited && len(beforeTestGoroutines) == 0 {
-		for _, g := range interestingGoroutines() {
-			beforeTestGoroutines[g] = true
-		}
-	}
-
-	for str := range beforeTestGoroutines {
-		logutil.BgLogger().Info("checkLeakAfterTest", zap.String("L163", str))
-	}
-
-	cnt := defaultCheckCnt
-	return func() {
-		defer func() {
-			beforeTestGoroutines = map[string]bool{}
-			testGoroutinesInited = false
-		}()
-
-		var leaked []string
-		for i := 0; i < cnt; i++ {
-			leaked = leaked[:0]
-			for _, g := range interestingGoroutines() {
-				if !beforeTestGoroutines[g] {
-					leaked = append(leaked, g)
-					for str := range beforeTestGoroutines {
-						logutil.BgLogger().Info("checkLeakAfterTest", zap.String("L180", str), zap.Int("i", i))
-					}
-					for _, str := range interestingGoroutines() {
-						logutil.BgLogger().Info("checkLeakAfterTest", zap.String("L183", str), zap.Int("i", i))
-					}
-				}
-			}
-			// Bad stuff found, but goroutines might just still be
-			// shutting down, so give it some time.
-			if len(leaked) != 0 {
-				time.Sleep(50 * time.Millisecond)
-				continue
-			}
-
-			return
-		}
-		for _, g := range leaked {
-			errorFunc(cnt, g)
-		}
-	}
-}
-
-// AfterTestT is used after all the test cases is finished.
-func AfterTestT(t *testing.T) func() {
-	errorFunc := func(cnt int, g string) {
-		t.Errorf("Test %s check-count %d appears to have leaked: %v", t.Name(), cnt, g)
-	}
-	return checkLeakAfterTest(errorFunc)
-}
-
 func TestInfo(t *testing.T) {
 	if !unixSocketAvailable() {
 		return
 	}
-	for i, str := range interestingGoroutines() {
-		t.Logf("TestInfo: BeforeTest %d %s", i, str)
-	}
-	BeforeTest()
-	defer func() {
-		AfterTestT(t)()
-		for i, str := range interestingGoroutines() {
-			t.Logf("TestInfo: AfterTest %d %s", i, str)
-		}
-	}()
+	testleak.BeforeTest()
+	defer testleak.AfterTestT(t)()
 	ddlLease := 80 * time.Millisecond
 	s, err := mockstore.NewMockStore()
 	if err != nil {

--- a/expression/evaluator_test.go
+++ b/expression/evaluator_test.go
@@ -38,7 +38,7 @@ var _ = Suite(&testVectorizeSuite2{})
 
 func TestT(t *testing.T) {
 	testleak.BeforeTest()
-	defer testleak.AfterTestT(t)
+	defer testleak.AfterTestT(t)()
 
 	CustomVerboseFlag = true
 	*CustomParallelSuiteFlag = true

--- a/util/testleak/leaktest.go
+++ b/util/testleak/leaktest.go
@@ -77,6 +77,18 @@ var testGoroutinesInited bool
 // BeforeTest gets the current goroutines.
 // It's used for check.Suite.SetUpSuite() function.
 // Now it's only used in the tidb_test.go.
+// Note: it's not accurate, consider the following function:
+// func loop() {
+//   for {
+//     select {
+//       case <-ticker.C:
+//         DoSomething()
+//     }
+//   }
+// }
+// If this loop step into DoSomething() during BeforeTest(), the stack for this goroutine will contain DoSomething().
+// Then if this loop jump out of DoSomething during AfterTest(), the stack for for this goroutine will not contain DoSomething().
+// Resulting in false positive leak report.
 func BeforeTest() {
 	for _, g := range interestingGoroutines() {
 		beforeTestGoroutines[g] = true

--- a/util/testleak/leaktest.go
+++ b/util/testleak/leaktest.go
@@ -87,8 +87,8 @@ var testGoroutinesInited bool
 //   }
 // }
 // If this loop step into DoSomething() during BeforeTest(), the stack for this goroutine will contain DoSomething().
-// Then if this loop jump out of DoSomething during AfterTest(), the stack for for this goroutine will not contain DoSomething().
-// Resulting in false positive leak report.
+// Then if this loop jumps out of DoSomething during AfterTest(), the stack for this goroutine will not contain DoSomething().
+// Resulting in false-positive leak reports.
 func BeforeTest() {
 	for _, g := range interestingGoroutines() {
 		beforeTestGoroutines[g] = true

